lightning/src/routing/router.rs

```diff
@@ -144,26 +143,26 @@ impl<'a, S: Score<ScoreParams = SP>, SP: Sized> Score for ScorerAccountingForInF
 				..usage
 			};
 
-			self.scorer.channel_penalty_msat(short_channel_id, source, target, usage, score_params)
+			self.scorer.read_lock().channel_penalty_msat(short_channel_id, source, target, usage, score_params)
```

Comment:
cool. lets split it *dance*

in terms of trait definitions, how do u see it play out?
```rust
pub trait Score $(: $supertrait)* {
	/// Handles updating channel penalties after failing to route through a channel.
	fn payment_path_failed(&mut self, path: &Path, short_channel_id: u64);

	/// Handles updating channel penalties after successfully routing along a path.
	fn payment_path_successful(&mut self, path: &Path);

	/// Handles updating channel penalties after a probe over the given path failed.
	fn probe_failed(&mut self, path: &Path, short_channel_id: u64);

	/// Handles updating channel penalties after a probe over the given path succeeded.
	fn probe_successful(&mut self, path: &Path);
}

pub trait ScoreUpdate $(: $supertrait)* {
	/// A configurable type which should contain various passed-in parameters for configuring the scorer,
	/// on a per-routefinding-call basis through to the scorer methods,
	/// which are used to determine the parameters for the suitability of channels for use.
	type ScoreParams;
	/// Returns the fee in msats willing to be paid to avoid routing `send_amt_msat` through the
	/// given channel in the direction from `source` to `target`.
	///
	/// The channel's capacity (less any other MPP parts that are also being considered for use in
	/// the same payment) is given by `capacity_msat`. It may be determined from various sources
	/// such as a chain data, network gossip, or invoice hints. For invoice hints, a capacity near
	/// [`u64::max_value`] is given to indicate sufficient capacity for the invoice's full amount.
	/// Thus, implementations should be overflow-safe.
	fn channel_penalty_msat(
		&self, short_channel_id: u64, source: &NodeId, target: &NodeId, usage: ChannelUsage, score_params: &Self::ScoreParams
	) -> u64;
}

```

Reply:
IIRC, when I played around with this awhile ago, I used the names `Score` and `ScoreLookUp`. So naming it the other way where the mutating operations define scoring a channel. Admittedly, the name `Score` is a bit ambiguous as in the first it is meant to be a verb while the second it is meant to be a noun. At very least, I think this version is a bit more intuitive absent a better alternative.

Reply:
sounds good!
after splitting the trait into two traits as the following 
```rust
pub trait ScoreUpdate $(: $supertrait)* {
	/// Handles updating channel penalties after failing to route through a channel.
	fn payment_path_failed(&mut self, path: &Path, short_channel_id: u64);

	/// Handles updating channel penalties after successfully routing along a path.
	fn payment_path_successful(&mut self, path: &Path);

	/// Handles updating channel penalties after a probe over the given path failed.
	fn probe_failed(&mut self, path: &Path, short_channel_id: u64);

	/// Handles updating channel penalties after a probe over the given path succeeded.
	fn probe_successful(&mut self, path: &Path);
}

pub trait ScoreLookUp $(: $supertrait)* {
	/// A configurable type which should contain various passed-in parameters for configuring the scorer,
	/// on a per-routefinding-call basis through to the scorer methods,
	/// which are used to determine the parameters for the suitability of channels for use.
	type ScoreParams;
	/// Returns the fee in msats willing to be paid to avoid routing `send_amt_msat` through the
	/// given channel in the direction from `source` to `target`.
	///
	/// The channel's capacity (less any other MPP parts that are also being considered for use in
	/// the same payment) is given by `capacity_msat`. It may be determined from various sources
	/// such as a chain data, network gossip, or invoice hints. For invoice hints, a capacity near
	/// [`u64::max_value`] is given to indicate sufficient capacity for the invoice's full amount.
	/// Thus, implementations should be overflow-safe.
	fn channel_penalty_msat(
		&self, short_channel_id: u64, source: &NodeId, target: &NodeId, usage: ChannelUsage, score_params: &Self::ScoreParams
	) -> u64;
}
```

the `LockableScore` wil look like this:
```rust
pub trait LockableScore<'a> {
	/// The [`Score`] type.
	type ScoreUpdate: 'a + ScoreUpdate;
	type ScoreLookUp: 'a + ScoreLookUp;

	/// The Write locked [`Score`] type.
	type WriteLocked: DerefMut<Target = Self::ScoreUpdate> + Sized;

	/// The Read locked [`Score`] type.
	type ReadLocked: Deref<Target = Self::ScoreLockUp> + Sized;

	/// Returns read locked scorer.
	fn read_lock(&'a self) -> Self::ReadLocked;

	/// Returns write locked scorer.
	fn write_lock(&'a self) -> Self::WriteLocked;
}
```
then when I want to change the implementations, I need up with something as the following:
```rust
impl<'a, T: 'a + ScoreLookUp, N: 'a + ScoreUpdate> LockableScore<'a> for Mutex<T> {
	type ScoreLookUp = T;
	type ScoreUpdate = N;
	type WriteLocked = MutexGuard<'a, Self::ScoreUpdate>;
	type ReadLocked = MutexGuard<'a, Self::ScoreLookUp>;

	fn read_lock(&'a self) -> Self::ReadLocked {
		Mutex::lock(self).unwrap()
	}

	fn write_lock(&'a self) -> Self::WriteLocked {
		Mutex::lock(self).unwrap()
	}
}
```
Im not sure about this one, the Mutex< T > doesnt seem correct.

In general does that look like the correct path forward?

Reply:
thanks @TheBlueMatt . 
whats the best way to approach  `ScorerAccountingForInFlightHtlcs`? I added new generic(`Su`) but not sure about that. now we implement `ScoreUpdate` and `ScoreLookUp` for `ScorerAccountingForInFlightHtlcs`  and I wonder whats the clearest way to write the `impl`  signature without too many generics. 
```rust
pub struct ScorerAccountingForInFlightHtlcs<'a, SP: Sized, Sc: 'a + ScoreUpdate , Su: 'a + ScoreLookUp<ScoreParams = SP>, S: LockableScore<'a, ScoreLookUp = Su, ScoreUpdate = Sc> + ?Sized> {
	scorer: &'a S,
	// Maps a channel's short channel id and its direction to the liquidity used up.
	inflight_htlcs: &'a InFlightHtlcs,
}
```

Reply:
pushed a [commit](https://github.com/lightningdevkit/rust-lightning/pull/2197/commits/0337df9caca80430bf01d32297d4064c510e54ab)  with some code changes. please lmk what u think. 
I need to update the docs so will look into that.